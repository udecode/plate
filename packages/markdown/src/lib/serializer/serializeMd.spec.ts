import { createTestEditor } from '../__tests__/createTestEditor';
import { testValue } from '../__tests__/testValue';
import { serializeMd } from './serializeMd';
const editor = createTestEditor();

describe('serializeMd', () => {
  it('should serialize a simple paragraph', () => {
    const result = serializeMd(editor as any, { value: testValue });
    expect(result).toMatchSnapshot();
  });

  describe('fixures', () => {
    // https://github.com/inokawa/remark-slate-transformer/issues/44
    it('should serialize marks with trailing whitespace', () => {
      const slateNodes = [
        {
          children: [
            {
              text: 'Make text ',
            },
            {
              bold: true,
              text: 'bold',
            },
            {
              text: ', ',
            },
            {
              italic: true,
              text: 'italic',
            },
            {
              text: ', ',
            },
            {
              text: 'underlined',
              underline: true,
            },
            {
              text: ', or apply a ',
            },
            {
              bold: true,
              italic: true,
              text: 'combination',
              underline: true,
            },
            {
              text: ' of these styles for a visually striking effect.',
            },
          ],
          type: 'p',
        },
      ];
      const result = serializeMd(editor as any, { value: slateNodes });
      expect(result).toMatchSnapshot();
    });

    // https://github.com/inokawa/remark-slate-transformer/issues/90
    it('Combination of bold and italic at the start of a block losing formatting', () => {
      const slateNodes = [
        {
          children: [
            {
              bold: true,
              italic: true,
              text: 'Italic',
            },
            {
              bold: true,
              text: ' in a bold paragraph',
            },
          ],
          type: 'p',
        },
        {
          children: [
            {
              bold: true,
              text: 'This is an ',
            },
            {
              bold: true,
              italic: true,
              text: 'Italic',
            },
            {
              bold: true,
              text: ' in a bold paragraph',
            },
          ],
          type: 'p',
        },
      ];

      const result = serializeMd(editor as any, { value: slateNodes });
      expect(result).toMatchSnapshot();
    });

    // https://github.com/inokawa/remark-slate-transformer/issues/145
    it('should serialize a value with inline code', () => {
      const slateNodes = [
        {
          children: [
            { bold: true, code: true, italic: true, text: 'Inline code' },
          ],
          type: 'p',
        },
        {
          children: [{ code: true, italic: true, text: 'Inline code' }],
          type: 'p',
        },
      ];

      expect(
        serializeMd(editor as any, { value: slateNodes })
      ).toMatchSnapshot();
    });

    it('should serialize a value within a block qoute as a single line', () => {
      const slateNodes = [
        {
          children: [
            { text: 'Block quote' },
            { text: ' with a new line ' },
            { bold: true, code: true, italic: true, text: ' Inline code' },
          ],
          type: 'blockquote',
        },
      ];

      expect(
        serializeMd(editor as any, { value: slateNodes })
      ).toMatchSnapshot();
    });

    it(
      String.raw`should serialize a \n within a block qoute as new line`,
      () => {
        const slateNodes = [
          {
            children: [
              { text: 'Block quote' },
              { text: '\n' },
              { text: 'with a new line' },
            ],
            type: 'blockquote',
          },
        ];

        expect(
          serializeMd(editor as any, { value: slateNodes })
        ).toMatchSnapshot();
      }
    );

    it(
      String.raw`should serialize two \n within a block qoute as two new lines`,
      () => {
        const slateNodes = [
          {
            children: [
              { text: 'Block quote' },
              { text: '\n' },
              { text: '\n' },
              { text: 'with a new line' },
            ],
            type: 'blockquote',
          },
        ];

        expect(
          serializeMd(editor as any, { value: slateNodes })
        ).toMatchSnapshot();
      }
    );

    it(
      String.raw`should serialize two leading \n at the end of a block qoute as a new line`,
      () => {
        const slateNodes = [
          {
            children: [{ text: 'Block quote' }, { text: '\n' }, { text: '\n' }],
            type: 'blockquote',
          },
        ];

        expect(
          serializeMd(editor as any, { value: slateNodes })
        ).toMatchSnapshot();
      }
    );

    it(
      String.raw`should serialize three leading \n at the end of a paragraph qoute as a new line`,
      () => {
        const slateNodes = [
          {
            children: [
              { text: 'Paragaph with two new Lines' },
              { text: '\n' },
              { text: '\n' },
              { text: '\n' },
            ],
            type: 'p',
          },
        ];

        expect(
          serializeMd(editor as any, { value: slateNodes })
        ).toMatchSnapshot();
      }
    );
  });

  it(
    String.raw`should serialize three leading \n at the end of a paragraph (inline) qoute as a new line`,
    () => {
      const slateNodes = [
        {
          children: [
            { text: 'Paragaph with two new Lines' + '\n' + '\n' + '\n' },
          ],
          type: 'p',
        },
      ];

      expect(
        serializeMd(editor as any, { value: slateNodes })
      ).toMatchSnapshot();
    }
  );

  it(
    String.raw`should serialize the leading break at the end of a block qoute as a <br />`,
    () => {
      const slateNodes = [
        {
          children: [{ text: 'Paragaph with a new Line' }, { text: '\n' }],
          type: 'p',
        },
      ];

      expect(
        serializeMd(editor as any, { value: slateNodes })
      ).toMatchSnapshot();
    }
  );

  it(String.raw`should serialize an empty paragraph to a <br />`, () => {
    const slateNodes = [
      {
        children: [{ text: '' }],
        type: 'p',
      },
      {
        children: [{ text: '' }],
        type: 'p',
      },
    ];

    expect(serializeMd(editor as any, { value: slateNodes })).toMatchSnapshot();
  });

  it(
    String.raw`should serialize paragraphs with only a new line to a <p><br /></p>`,
    () => {
      const slateNodes = [
        {
          children: [{ text: '\n' }],
          type: 'p',
        },
      ];

      expect(
        serializeMd(editor as any, { value: slateNodes })
      ).toMatchSnapshot();
    }
  );

  it(
    String.raw`should serialize paragraphs with two a new line to a \\n<br />`,
    () => {
      const slateNodes = [
        {
          children: [{ text: '\n' }, { text: '\n' }],
          type: 'p',
        },
      ];

      expect(
        serializeMd(editor as any, { value: slateNodes })
      ).toMatchSnapshot();
    }
  );
});
