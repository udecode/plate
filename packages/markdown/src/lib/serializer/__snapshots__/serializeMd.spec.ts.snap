// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializeMd fixures Combination of bold and italic at the start of a block losing formatting 1`] = `
"***Italic*** **in a bold paragraph**

**This is an *Italic*** **in a bold paragraph**
"
`;

exports[`serializeMd fixures should serialize a \\n within a block qoute as new line 1`] = `
"> Block quote\\
> with a new line
"
`;

exports[`serializeMd fixures should serialize a value with inline code 1`] = `
"***\`Inline code\`***

*\`Inline code\`*
"
`;

exports[`serializeMd fixures should serialize a value within a block qoute as a single line 1`] = `
"> Block quote with a new line ***\` Inline code\`***
"
`;

exports[`serializeMd fixures should serialize marks with trailing whitespace 1`] = `
"Make text **bold**, *italic*, <u>underlined</u>, or apply a ***<u>combination</u>*** of these styles for a visually striking effect.
"
`;

exports[`serializeMd fixures should serialize three leading \\n at the end of a paragraph qoute as a new line 1`] = `
"Paragaph with two new Lines\\
\\ 
<br />
"
`;

exports[`serializeMd fixures should serialize two \\n within a block qoute as two new lines 1`] = `
"> Block quote\\
> \\
> with a new line
"
`;

exports[`serializeMd fixures should serialize two leading \\n at the end of a block qoute as a new line 1`] = `
"> Block quote\\ 
> <br />
"
`;

exports[`serializeMd should serialize a simple paragraph 1`] = `
"# Playground

## Comments and suggestions

Add comments to your content to provide additional conte1xt, , or collaborate with others

### Heading

* 1

  * 2

    * 3

      * 4

| 1-1 | 1-2 |
| --- | --- |
| 2-1 | 2-2 |

> blockQuote

***

normal **bold***italic*~~strikethrough~~\`code\`***combine***[link](https://example.com)$E=mc^2$

![test](https://images.unsplash.com/photo-1712688930249-98e1963af7bd?q=80\\&w=2070\\&auto=format\\&fit=crop\\&ixlib=rb-4.0.3\\&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D "test")

\`\`\`javascript
// Use code blocks to showcase code snippets
function greet() {
  console.info('Hello World!');
}
\`\`\`

$$
f(x)
$$

<toc />

* [x] Todo item

* Numbered list item

Centered text

Text with custom line height

Mention: @User

Date: 2024-01-01

Text with <highlight style="background-color: yellow;">highlight</highlight> and <u>underline</u> and keyboard formatting

Text with color and background color
"
`;

exports[`serializeMd should serialize an empty paragraph to a <br /> 1`] = `
"<br />

<br />
"
`;

exports[`serializeMd should serialize paragraphs with only a new line to a <p><br /></p> 1`] = `
"<p><br /></p>
"
`;

exports[`serializeMd should serialize paragraphs with two a new line to a \\\\n<br /> 1`] = `
"\\ 
<br />
"
`;

exports[`serializeMd should serialize the leading break at the end of a block qoute as a <br /> 1`] = `
"Paragaph with a new Line
<br />
"
`;
